Code/Framework/AzCore/AzCore/IO/Streamer/RecentlyUsedIndex.h

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#pragma once
+
+#include <AzCore/std/functional.h>
+#include <AzCore/std/numeric.h>
+#include <AzCore/std/smart_ptr/unique_ptr.h>
+#include <AzCore/std/typetraits/is_integral.h>
+
+namespace AZ::IO
+{
+    //! Uses an optimized double linked list to keep track of the most and least recently used index.
+    template<typename T>
+    class RecentlyUsedIndex
+    {
+        static_assert(
+            AZStd::is_integral_v<T> && AZStd::is_unsigned_v<T>,
+            "Recently Used Index only supports unsigned integers as they need to represent an index.");
+
+    public:
+        static constexpr T InvalidIndex = AZStd::numeric_limits<T>::max();
+
+        RecentlyUsedIndex() = default;
```

Comment:
Should the RecentlyUsedIndex expose a default constructor?

Default constructing a RecentlyUsedIndex entry followed by invoking FlushAll would cause a nullptr defererence of the `m_indices_previous` member.

Reply:
Unfortunately this is needed because some of the `*StorageDrive`s use lazy initialization so they only take up memory if they're actually used. This includes the uses of the `RecentlyUsedIndex`.